lib/routemaster/api_client.rb

```diff
@@ -25,16 +25,23 @@
 
 module Routemaster
   class APIClient
+
+    # Memoize the root resources at Class level so that we don't hit the cache
+    # all the time to fetch the root resource before doing anything else.
+    @@root_resources = {}
+
     def initialize(middlewares: [],
                    listener: nil,
                    response_class: nil,
                    metrics_client: nil,
                    source_peer: nil)
-      @listener = listener
-      @middlewares = middlewares
+
+      @listener       = listener
+      @middlewares    = middlewares
       @response_class = response_class
       @metrics_client = metrics_client
-      @source_peer = source_peer
+      @source_peer    = source_peer
+      @lock           = Mutex.new
```

Comment:
Your lock is per `APIClient` instance, so it doesn't actually protect if it's instantiated more than once.
I'd suggest this classic pattern:

```ruby
@@root_resources = {}.extend(MonitorMixin)

# ...

def discover(url)
  @@root_resources.synchronize do
    @@root_resources[url] = get(url)
  end
end
```

In this case though, your data is write-only, so could avoid the lock altogether, ie the trivial implementation would work:

```ruby
def discover(url)
  @@root_resources[url] ||= get(url)
end
```

... at the risk of making the request twice initially.